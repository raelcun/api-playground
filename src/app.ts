import Koa from 'koa'
import bodyParser from 'koa-bodyparser'
import helmet from 'koa-helmet'

import { getEmergencyLogger } from '@lib/emergency-logger'
import { logRequest } from '@modules/logging-middleware'
import { router } from '@root/routes'

const app = new Koa()

app.on('error', (err, ctx) => {
  getEmergencyLogger().error('koa logged error', { err, ctx })
})

app.use(helmet())
app.use(bodyParser())
app.use(logRequest)
app.use(router.routes())
app.use(router.allowedMethods())

export { app }
