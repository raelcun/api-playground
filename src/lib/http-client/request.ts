import { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios'
import { taskEither as TE } from 'fp-ts'
import { pipe } from 'fp-ts/lib/function'

import { Err } from '@lib/error'
import { logErrorsTE, LoggerProvider } from '@lib/logger'
import { tryCatchTE } from '@lib/utils'

import { BaseRequestClient } from './types'

export const request = (instance: BaseRequestClient) => <T>(
  config: AxiosRequestConfig,
  getLogger: LoggerProvider,
): TE.TaskEither<Err, AxiosResponse<T>> => {
  getLogger().log({
    level: 'trace',
    payload: { message: 'making http request', meta: { config } },
  })

  return pipe(
    tryCatchTE<Err, AxiosResponse<T>>(
      () => instance.request<T>(config),
      (e: AxiosError | Err): Err => ({
        code: e.code || 'UNKNOWN_HTTP_ERROR',
        message: e.message,
      }),
    ),
    logErrorsTE(getLogger(), 'error'),
  )
}
