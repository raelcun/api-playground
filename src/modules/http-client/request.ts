import { TaskEither, tryCatch } from 'fp-ts/lib/TaskEither'
import { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios'
import { BaseRequestClient, RequestError } from './types'

export const request = (instance: BaseRequestClient) => <T>(
  config: AxiosRequestConfig,
): TaskEither<RequestError, AxiosResponse<T>> =>
  tryCatch<RequestError, AxiosResponse<T>>(
    (): Promise<AxiosResponse<T>> => instance.request<T>(config),
    (e: AxiosError): RequestError =>
      e.code === 'ECONNABORTED' ? { code: 'TIMEOUT' } : { code: 'NETWORK_ERROR' },
  )
