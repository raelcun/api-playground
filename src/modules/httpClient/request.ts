import { taskEither as TE } from 'fp-ts'
import { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios'
import { BaseRequestClient, RequestError } from './types'

export const request = (instance: BaseRequestClient) => <T>(
  config: AxiosRequestConfig,
): TE.TaskEither<RequestError, AxiosResponse<T>> =>
  TE.tryCatch<RequestError, AxiosResponse<T>>(
    (): Promise<AxiosResponse<T>> => instance.request<T>(config),
    (e: AxiosError): RequestError =>
      e.code === 'ECONNABORTED' ? { code: 'TIMEOUT' } : { code: 'NETWORK_ERROR' },
  )
