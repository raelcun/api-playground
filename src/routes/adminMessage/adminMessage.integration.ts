import HttpStatus from 'http-status-codes'
import request from 'supertest'

import { app } from '@root/app'

describe('adminMessage', () => {
  test('should return body when authorized', async () => {
    await request(app.callback())
      .post('/v1/adminMessage')
      .send({
        role: 'admin',
        message: 'foobar',
      })
      .expect(HttpStatus.OK, 'foobar')
  })

  test('should reject unauthorized request ', async () => {
    await request(app.callback())
      .post('/v1/adminMessage')
      .send({
        role: 'user',
        message: 'foobar',
      })
      .expect(HttpStatus.UNAUTHORIZED, 'Unauthorized')
  })

  test('should validate post body', async () => {
    await request(app.callback())
      .post('/v1/adminMessage')
      .send({ role: 'admin' })
      .expect(HttpStatus.BAD_REQUEST, {
        code: 'BODY_VALIDATION_ERROR',
        message: 'Expecting string at message but instead got: undefined.',
      })
  })
})
