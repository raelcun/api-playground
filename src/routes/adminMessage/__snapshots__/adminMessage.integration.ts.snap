// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adminMessage should return body when authorized 1`] = `
Object {
  "apiName": "api-playground",
  "apiVersion": Any<String>,
  "data": "foobar",
  "id": Any<String>,
}
`;

exports[`adminMessage should validate post body 1`] = `
Object {
  "apiName": "api-playground",
  "apiVersion": Any<String>,
  "error": Object {
    "code": "BODY_VALIDATION_ERROR",
    "message": "Expecting string at message but instead got: undefined",
  },
  "id": Any<String>,
}
`;
